src/foam/nanos/ruler/Rule.js

```diff
@@ -168,13 +169,24 @@
         {
           name: 'oldObj',
           type: 'FObject'
+        },
+        {
+          name: 'isAsync',
+          type: 'Boolean'
         }
       ],
       javaCode: `
-        return getEnabled()
-          && getPredicate().f(
-            x.put("NEW", obj).put("OLD", oldObj)
-          );
+        Boolean isAvailable = null != (isAsync ? getAsyncAction() : getAction());
```

Comment:
Why do we combine these two different checks in one? Shouldn't we separate logic for rule.predicate and action == null? If we were to check if a rule is applicable it will be impossible to tell whether predicate is false or action is null

Reply:
Yeah, that makes sense to not combine them. Also, we can check if action != null before checking if rule is applicable so we don't evaluate the predicate at all if it is the case.